pkg/k8s/completion.go

```diff
@@ -0,0 +1,171 @@
+//
+// Complete is generic enough so that it can find suggestions for any type of resource
+// in a Kubernetes cluster. It does this by first querying what GroupVersion a resource
+// belongs to and then does a dynamic `List` query to get resources under that GroupVersion
+func (c *CommandCompletion) Complete(args []string, toComplete string) ([]string, error) {
+	ctx, cancelFn := context.WithTimeout(context.Background(), 5*time.Second)
+	defer cancelFn()
+
+	suggestions := []string{}
+	if len(args) == 0 && toComplete == "" {
+		return StatAllResourceTypes, nil
+	}
+
+	if len(args) == 0 && toComplete != "" {
+		for _, t := range StatAllResourceTypes {
+			if strings.HasPrefix(t, toComplete) {
+				suggestions = append(suggestions, t)
+			}
+		}
+		return suggestions, nil
+	}
+
+	resType, err := CanonicalResourceNameFromFriendlyName(args[0])
+	if err != nil {
+		return nil, fmt.Errorf("%s not a valid resource name", args)
+	}
+
+	// if we are looking for namespace suggestions clear namespace selector
+	if resType == "namespace" {
+		c.namespace = ""
+	}
+
+	// Similar to kubectl, we don't provide resource completion
+	// when the resource provided is in format <kind>/<resourceName>
+	if strings.Contains(args[0], "/") {
+		return []string{}, nil
+	}
+
+	gvr, err := c.getGroupVersionKindForResource(resType)
+	if err != nil {
+		return nil, err
+	}
+
+	uList, err := c.k8sAPI.DynamicClient.
+		Resource(*gvr).
+		Namespace(c.namespace).
+		List(ctx, metav1.ListOptions{})
+	if err != nil {
+		return nil, err
+	}
+
+	otherResources := args[1:]
+	for _, u := range uList.Items {
+		name := u.GetName()
+		if strings.HasPrefix(name, toComplete) &&
+			!containsResource(name, otherResources) {
+			suggestions = append(suggestions, name)
+		}
+	}
+
+	return suggestions, nil
+}
+
+func (c *CommandCompletion) getGroupVersionKindForResource(resourceName string) (*schema.GroupVersionResource, error) {
+	_, apiResourceList, err := c.k8sAPI.Discovery().ServerGroupsAndResources()
+	if err != nil {
+		return nil, err
+	}
+
+	// find the plural name to ensure the resource we are searching for is not a subresource
+	// i.e. deployment/scale
+	pluralResourceName, err := PluralResourceNameFromFriendlyName(resourceName)
+	if err != nil {
+		return nil, fmt.Errorf("%s not a valid resource name", resourceName)
+	}
+
+	var gvr *schema.GroupVersionResource
+	for _, res := range apiResourceList {
+		for _, r := range res.APIResources {
+			if strings.ToLower(r.Kind) == resourceName && r.Name == pluralResourceName {
+				gv := strings.Split(res.GroupVersion, "/")
+
+				if len(gv) == 1 && gv[0] == "v1" {
+					gvr = &schema.GroupVersionResource{
+						Version:  gv[0],
+						Resource: r.Name,
+					}
+					break
+				}
+
+				if len(gv) != 2 {
+					return nil, fmt.Errorf("could not find the requested resource")
+				}
+
+				gvr = &schema.GroupVersionResource{
+					Group:    gv[0],
+					Version:  gv[1],
+					Resource: r.Name,
+				}
```

Comment:
qq: would we need a break here to once we find the gvr? do we keep looping if we don't break?

Reply:
This makes me think that this logic needs to be in a separate function. The multiple breaks in this code sections maybe be a little difficult to read through in the first pass. If we put this in a function, we can just immediately return the `GroupVersionResource` when we encounter it and avoid having multiple `break` statements to exist the nested `for` loop